File: solutions/剑指 Offer 04. 二维数组中的查找 LCOF/solution2.py
class Solution:
    def findNumberIn2DArray(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        n, m = len(matrix), len(matrix[0])
        for i in range(n):
            for j in range(m-1, -1, -1):
                if matrix[i][j] == target:
                    return True
                elif matrix[i][j] > target:
                    m -= 1
                elif matrix[i][j] < target:
                    break
        return False    

File: solutions/剑指 Offer 53 - I. 在排序数组中查找数字 I/solution2.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target, bool lower_bound) {
        int n = nums.size(), left = 0, right = n, mid;
        while (left < right) {
            mid = (left + right) / 2;
            if (nums[mid] == target) {
                if (lower_bound) right = mid;
                else left = mid + 1;
            }
            else if (nums[mid] > target) right = mid;
            else left = mid + 1;
        }
        return lower_bound ? left : right - 1;
    }
    int search(vector<int>& nums, int target) {
        int left_idx = binarySearch(nums, target, true);
        int right_idx = binarySearch(nums, target, false);
        return right_idx - left_idx + 1;
    }
};